### ISSUE TITLE
Parser should accept `echo {foo`

### ISSUE THREAD

**Issue by lilyball** [fish-shell/fish-shell#5880]
Current `master` no longer expands braces that do not contain a comma or variable reference (see 967c1d51eec0f840273c20a326152ea72f920d3d). This means that e.g. `echo {foo}` echoes `{foo}`.

As such, we should probably allow for expressions like `echo {foo`, which should echo `{foo`, as the brace is not considered special. The expression `echo {foo,` would continue to error out, as this is an attempted parameter expansion.

According to [this comment](https://github.com/fish-shell/fish-shell/issues/5869#issuecomment-493741430) the issue with accepting `echo {foo` is that the tokenizer checks matching braces (and it references #2485).

**Comment by mqudsi** [comment]
I guess it depends on whether you consider a sole opening brace to be special or not. How often do you have to type a literal `{` but not a literal `}` at any point thereafter? It seems like it would be an incredibly small number of commands that would benefit by this, to the extent that they may be overshadowed by typos where brace expansion was intended?

**Comment by lilyball** [comment]
The point here is that as of 967c1d51eec0f840273c20a326152ea72f920d3d `echo {foo}` no longer treats the braces as special, and since they're not special, then `echo {foo` should also work. Braces only become special once they contain a comma or variable reference, so `echo {foo,` would become an error, but up until you type the comma the brace should be fine.

**Comment by mqudsi** [comment]
Yes, I understand that. @faho's 967c1d5 was really more of special-casing an empty `{}` to work around a common user error rather than actually changing fish's syntax. My earlier reply was questioning the utility of taking that further.

**Comment by lilyball** [comment]
We special-cased `{}` a while ago. 967c1d51eec0f840273c20a326152ea72f920d3d is actually a deliberate change to make braces non-special unless they contain commas or variables, largely motivated by the fact that Git uses braces for some of its syntax (e.g. `master@{upstream}` to reference the upstream branch for master, or `HEAD@{2}` to reference something in the reflog, or `v0.3.2^{commit}` to convert a tag into the pointed-to commit, etc). Also see #5869 for discussion.

**Comment by Phidica** [comment]
I agree with @mqudsi that allowing a sole `{` to be treated literally would surely have little practical benefit, and could more often trip people up by *not* throwing a syntax error, because most likely they have forgotten the closing brace. 

Consider how we would describe the brackets rules for fish: "the round brackets `( )` always have to come in pairs, the square brackets `[ ]` always have to come in pairs, the braces `{ }` always have to come in pairs *if* you want to perform *certain* tasks with them"...? Why not just stick to the pattern and say that braces always *always* have to come in pairs?

Additionally, I think it's worth noting that braces without a comma or variable expansion in them *are* still a little bit special, because control characters (`! & | < >`) are treated as literals if they appear inside the braces, and extra spaces get eaten up (only from the front though...which is a regression from 3.0). So in 3.1, `echo { foo ;` is invalid, while `echo { foo ; }` is valid and prints `{foo ; }`. If we made this change, then `echo { foo ; }` would probably print `{ foo` but fail because `}` isn't a valid command name. I grant this is an extreme example and unlike what any user is probably doing with braces, but I find looking at edge cases to be illuminating on what we feel the general behaviour should (or shouldn't) be.

To me, it makes more sense that leaving the braces unbalanced just isn't allowed.

**Comment by lilyball** [comment]
I can't think of any other case where a single job comprised entirely of literal text becomes invalid if you delete a character off the end.

And besides, square brackets don't always have to come in pairs. I'm not actually sure why `echo foo[bar` isn't legal, I didn't realize that and think it should be changed, but `echo [foo` is perfectly legal.

**Comment by Phidica** [comment]
How about `[ 1 -eq 1 ]`? It is invalid without the closing `]`. And indeed it's another case where brackets must be balanced despite that not serving any purpose.

Something else to consider: should it also be the case that a sole `}` is treated as a literal character?
```fish
$ echo foo}
foo}
```
Because, again, I think it's useful to consider unmatched braces to be errors for usability when the reality is that the matching brace was forgotten.

**Comment by lilyball** [comment]
`[ 1 -eq 1 ]` isn't literal text though, it's a different syntactic structure entirely.

> Something else to consider: should it also be the case that a sole `}` is treated as a literal character?

I would say yes. Braces should be treated just like other literal text unless they contain a comma or variable expansion. A lone `}` can't contain anything.

I wrote in the issue text that `echo {foo,` should still be an error as that's attempted parameter expansion. At this point I'm on the fence and could be swayed either way about that. But `echo {foo` and `echo foo}` should both be legal.

**Comment by mqudsi** [comment]
I think correctness is more important than terseness or flexibility. The cost of getting a parser error and then having to escape your input is dwarfed by the cost of having your mistyped input parsed as something else.

**Comment by ridiculousfish** [comment]
This is not simple to implement and should not block 3.2.0; pushing to 3.3.0.

**Comment by ChristoferK** [comment]
> Why not just stick to the pattern and say that braces always _always_ have to come in pairs?

A better option would be to say that none of `(`, `)`, `{`, `}`, `[`, `]` need to come in pairs, and are assumed to be literals until paired.

A warning message printed to `stderr` (and optionally turned off) could be a helpful reminder to those who might be likely to unintentionally leave these sorts of things unpaired.

**Comment by faho** [comment]
> A better option would be to say that none of (, )

Definitely not `(,)` - those are command substitutions and fairly important, and *should* come in pairs.

`{}` (braces that don't trigger brace expansion don't need to be paired) is this, `[]` (brackets that don't trigger indexing don't need to be paired) is #2485.

**Comment by ChristoferK** [comment]
> Definitely not `(,)` - those are command substitutions and fairly important, and _should_ come in pairs.

Accepted.
 
> `[]` (brackets that don't trigger indexing don't need to be paired) is #2485.

I'm trying to decipher what the outcome is there, although the issue dates back to 2015, so my guess is that the decision was that `[` and `]` shouldn't be used unpaired unless escaped or quoted given that the parser errs on most of those situations...?
